### DISCUSSION TITLE
Add Support for skip_final_snapshot in RDS clusters

### DISCUSSION THREAD

**Discussion by dorayakikun** [aquasecurity/trivy#5130] [Ideas]
## Abstract

I want to detect some resources which set `skip_final_snapshot = true` in terraform resources by trivy's custom policy.

## Motivation

When destroying an RDS Cluster with Terraform, the backup of the entire cluster is deleted even if the Backup function is enabled.
In the above cases, permanent data loss is risky if recreation is unintentionally performed. (*)

`*` Automatic backups on cluster deletion have recently been released, but these have not yet been implemented in the terraform aws provider.

https://aws.amazon.com/about-aws/whats-new/2023/08/amazon-aurora-automatic-backup-deleted-clusters/

To avoid this situation, I want to use rego to detect resources with `skip_final_snapshot = true`.

## Proposal

I'm considering creating the below codes.

https://github.com/dorayakikun/defsec/commit/bd9efe9e726d63f9b9f1dcbffa7a12e57f7f154c

## Concern

`skip_final_snapshot` is a particular property.
Adding implementations to AWS Adapters and CloudFormation Adapters makes it difficult.
(I'm concerned that it will always be set to `false`.)

I am concerned that this violates defsec's design policy.

If you have any other better ideas, please let me know!

### Target

AWS

### Scanner

Misconfiguration

**Comment by simar7** [comment]
Thank you for the discussion and the insight.

> I'm concerned that it will always be set to false

If I understand correctly, you want to see if the value is set to `true` yes? If so, majority of the situations will have it set to `false` which is the default?

If the above is correct, then I think it's reasonable to add this as a rule. However, if majority of them will be flagged as an issue, I think we will need to think more.

**Reply by dorayakikun** [reply]
@simar7 
Thank you for your reply!

> If I understand correctly, you want to see if the value is set to true yes?

yes

> If so, majority of the situations will have it set to false which is the default?

> If the above is correct, then I think it's reasonable to add this as a rule. 

I envisage that it's the majority.
Because this property's default value is `false`.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/rds_cluster#skip_final_snapshot

> (Optional) Determines whether a final DB snapshot is created before the DB cluster is deleted. If true is specified, no DB snapshot is created. If false is specified, a DB snapshot is created before the DB cluster is deleted, using the value from final_snapshot_identifier. Default is false.

In light of the above, would it be possible to make the modifications I have in mind?

**Reply by simar7** [reply]
OK understood. If the majority of the cases will be set to `false`, I think we can add this in to flag if it is set to `true`.

**Comment by simar7** [comment]
I created an issue to track it https://github.com/aquasecurity/trivy/issues/5147 - you can submit a PR against that issue. Thank you for your contribution.

**Reply by dorayakikun** [reply]
Thank you for your review!
I've created https://github.com/aquasecurity/defsec/pull/1449 .
